Log drop notification

nordic-krch opened the pull request:
Extended backend interface to allow notifying backend  that log messages has been dropped due to insufficient internal buffer size. Notification contains number of log messages dropped since last notification. It is optional for a backend to implement handler.

It touches #11763 for log core. Howeverm updates in backends are required to fix the issue.

nordic-krch committed:
tests: subsys: logging: Add test for message drop notification

Extended logger suite to test correctness of reporting dropped
log messages.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pfalcon commented:
Commit message typo:

> Howeverm

jarz-nordic commented on include/logging/log_backend.h:
```diff
@@ -94,7 +95,25 @@ static inline void log_backend_put(const struct log_backend *const backend,
+static inline void log_backend_dropped(const struct log_backend *const backend,
+				       u32_t cnt)
+{
+	assert(backend);
```
Please use `__ASSERT_NO_MSG`

nordic-krch replied:
done, created dedicated commit for cleanup (doxygen + asserts)

pfalcon commented:
> It is optional for a backend to implement handler.

That's exactly why we have tickets titled "logging cannot be trusted". Why after all this discussion not make it mandatory? Careless programmer can still stub it with an empty function.

pfalcon commented on include/logging/log_backend.h:
```diff
@@ -94,7 +95,25 @@ static inline void log_backend_put(const struct log_backend *const backend,
- * @brief Function for reconfiguring backend to panic mode.
+ * @brief Notify backend about dropped log messages.
+ *
+ * Function is optional.
```
To what exactly this "optional" note applies? This is docstring for log_backend_dropped(), it is not optional and should be called properly by the logging core.

pfalcon commented on include/logging/log_backend.h:
```diff
@@ -105,7 +124,7 @@ static inline void log_backend_panic(const struct log_backend *const backend)
 
 /**
- * @brief Function for setting backend id.
+ * @brief Set backend id.
```
All these docstring changes should have gone as a separate commit.

jakub-uC reviewed (changes requested):
Minor change needed

pfalcon commented on tests/subsys/logging/log_core/src/log_core_test.c:
```diff
@@ -25,6 +25,9 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 typedef void (*custom_put_callback_t)(struct log_backend const *const backend,
 				      struct log_msg *msg, size_t counter);
+
+static bool in_panic;
```
Commit says it's "Add test for message drop notification", but suddenly, "in_panic" is added here.

nordic-krch replied:
in the commit i've reordered tests to put panic test at the bottom. Unfortunately, there is no cleanup  between test cases (e.g. reset) so logger state stays between tests and there is no going back when panic mode is started. flag in_panic is precaution for the future if someone adds test case after panic test case.

nordic-krch replied:
ok, let me split that into 2 commits. first tests reordering and then adding new test cases.

pfalcon replied:
> ok, let me split that into 2 commits

Only if you find that not too cumbersome. And actually, perfectalize tests is not worth it, better to do it like that next time in such situation.

However, if you'd split commit to logging.h into docstring updates vs drop count support, I'd truly appreciate.

nordic-krch replied:
log_backend.h cleanup extracted to separate commit.

pfalcon commented on tests/subsys/logging/log_core/src/log_core_test.c:
```diff
@@ -288,34 +301,6 @@ static void test_log_arguments(void)
 		      "Unexpected amount of messages received by the backend.");
 }
 
-/*
```
Such massive removals unrelated to the commit topic are really not ok.

pfalcon reviewed (commented):
IMHO, some better structuring of changes would be helpful.

pfalcon commented:
> Commit message typo:

>>    Howeverm

Ok, that was PR description, not commit message.

pfalcon commented:
So, why not include a backend change in this PR?

nordic-krch commented:
> That's exactly why we have tickets titled "logging cannot be trusted". Why after all this discussion not make it mandatory? Careless programmer can still stub it with an empty function.

Reasoning for optional:
- to allow us to introduce this feature backend by backend
- backward compatibility (someone may already implemented log own log backend)
- some backends may not be interested in handling that (e.g. flash backend).

Main reason is the first one. Once backends are aligned we can make it obligatory like other calls if we can leave with new requirement for backend interface (api breaker).

nordic-krch commented:
@pfalcon 
> So, why not include a backend change in this PR?

I prefer to do it step by step. Bulk PRs tend to stuck.

pfalcon commented:
> I prefer to do it step by step. Bulk PRs tend to stuck.

That's true, and sounds good.

nordic-krch committed:
logging: Add dropped messages notification to backends

Extended backend interface to allow notifying backend
that log messages has been dropped due to insufficient
internal buffer size. Notification contains number of
log messages dropped since last notification. It
is optional for a backend to implement handler.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
logging: Clean up in log_backend.h (asserts and doxygen)

Cleaning up doxygen comments. Replacing assert() with __ASSERT_NO_MSG.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pfalcon reviewed (approved):
Looks good, thanks for updates, let's move along with this, +1.
